SAMOA-16: Add an adapter for Apache Flink-Streaming

senorcarbone opened the pull request:
This PR includes support for Apache Flink as an adapter. We haven't updated the documentation yet but we could perhaps do it in a follow up PR, since it seems you are already working on a docs redesign.

gdfm commented:
Thanks. There seem to be some errors in the tests (Kryo serialization for what I can see).
Probably some Kryo version mismatch. Could you have a look at them?

senorcarbone commented:
Sure, we are looking into it! Kryo version incompatibility seems to be trickier than we thought.

gdfm commented:
If we need to update Kryo on our side, we'll be happy to do so.

senorcarbone commented:
it seems that the earliest Kryo version that Flink is compatible with is v2.23.0. However, Storm tests seem to fail to initialise the serialiser for versions  greater than 2.17. I think the current Apache Storm version (v.0.10.0) supports newer Kryo versions so maybe a first approach would be to try upgrade Storm first. What do you think?

gdfm commented:
Makes sense, as we want to update the Storm dependency anyway.
However, it will for sure happen to have conflicting dependencies on different platforms, and we don't want to mix them. The tests should ensure this isolation, i.e., never mix two platforms, and the different profiles aim at that.

We have a single kryo.version variable to ensure that we don't use different Kryo versions around the codebase, but in this case it seems to me that it is necessary.
If Flink depends on 2.23, but we have it set on 2.17, then we should be able to override it in the samoa-flink module. Do you agree?

senorcarbone commented:
Good point. I just overrided the kryo version in the flink build and tested it locally. Also, all module tests seem to be passing on travis.

arinto commented:
I can't build this PR successfully on my local. Here's the error message:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project samoa-flink: Compilation failure: Compilation failure:
[ERROR] /Users/arinto/git/incubator-samoa/samoa-flink/src/main/java/com/yahoo/labs/flink/topology/impl/FlinkEntranceProcessingItem.java:[54,79] <anonymous com.yahoo.labs.flink.topology.impl.FlinkEntranceProcessingItem$1> is not abstract and does not override abstract method cancel() in org.apache.flink.streaming.api.function.source.SourceFunction
[ERROR] /Users/arinto/git/incubator-samoa/samoa-flink/src/main/java/com/yahoo/labs/flink/topology/impl/FlinkEntranceProcessingItem.java:[64,25] method does not override or implement a method from a supertype
```

Apparently there are API changes <https://github.com/apache/flink/commit/8436e9ce31b52f1bd8c55b8e8c50cafb57cff84f> due to [FLINK-1625](https://issues.apache.org/jira/browse/FLINK-1625). Probably we can use stable version of Flink for this PR. What do you guys think?

senorcarbone commented:
hey Arinto! Thanks for looking into it. I agree, let's stick to the last stable release (0.8.1) for now. We will commit some appropriate changes for it today. cheers

senorcarbone commented:
We reviewed the changes after 0.8.1 and a lot of things have changed since then, many constructs such as the type extractors we are currently using will not be functioning in 0.8.1. Thus, we decided to just make a final patch for the first RC of 0.9.0 that is coming very soon, most probably end of next week. It looks like it is the best way to go :)

rmetzger commented:
+1
I think we need to release some kind of 0.9-preview release soon.

gdfm commented:
Hi,
if I understand correctly, you are waiting for Flink to release 0.9 before updating the PR, am I correct?

senorcarbone commented:
yes, we are planning a 0.9.0-milestone release so we can use that as the first supported stable version for the PR

senorcarbone commented:
We finally have an early Flink release milestone for this PR  (0.9.0-milestone-1)! I think it is a good point to review it whenever you can. We have tested several tasks form our side with this version (eg. prequential evaluation with vertical hoeffding trees) and it seems to be running accurately.

gdfm commented:
Thanks for the update and congrats on the almost-release :)
I will check it out in the weekend.

gdfm commented on bin/samoa-flink.properties:
```diff
@@ -0,0 +1,35 @@
+# #%L
+# SAMOA
+# %%
+# Copyright (C) 2013 Yahoo! Inc.
```
The copyright info is outdated.
I think this has recently been fixed in the pom file.

gdfm commented on samoa-api/build/test/data/dfs/data/data1/current/BP-62506341-193.10.66.77-1421760460444/current/VERSION:
```diff
@@ -0,0 +1,5 @@
+#Tue Jan 20 14:27:40 CET 2015
```
These files should not be checked in (actually, we should remove them automatically after the tests finish).

gdfm commented on samoa-api/src/main/java/com/yahoo/labs/samoa/streams/PrequentialSourceTopologyStarter.java:
```diff
@@ -0,0 +1,52 @@
+//package com.yahoo.labs.samoa.streams;
```
This class was removed.

gdfm commented on samoa-flink/src/main/java/com/yahoo/labs/flink/FlinkDoTask.java:
```diff
@@ -0,0 +1,129 @@
+		}
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		logger.debug("Creating the factory\n");
```
Do we need the final '\n' ?

gdfm commented on samoa-flink/src/main/java/com/yahoo/labs/flink/FlinkDoTask.java:
```diff
@@ -0,0 +1,129 @@
+		CircleDetection detCircles = new CircleDetection();
+		List<List<Integer>> circles = detCircles.getCircles(graph);
+
+		//update PIs, regarding being part fo a circle.
```
Typo 'of'

gdfm commented on samoa-flink/src/main/java/com/yahoo/labs/flink/com/yahoo/labs/flink/helpers/CircleDetection.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.List;
+import java.util.Stack;
+
+public class CircleDetection {
```
Could you add a comment explaining why we need circle detection?

gdfm commented on samoa-flink/src/main/java/com/yahoo/labs/flink/com/yahoo/labs/flink/helpers/Utils.java:
```diff
@@ -0,0 +1,69 @@
+
+public class Utils {
+
+	public enum Partitioning {SHUFFLE, ALL, GROUP}
```
Are these Partitioning Flink-specific or could we use the default ones in SAMOA?

senorcarbone replied:
You are right, we can. I replaced the usage of these with the default partitioning parameters of SAMOA several commits before, just forgot to remove their definition. I just removed it completely

gdfm commented on samoa-flink/src/main/java/com/yahoo/labs/flink/topology/impl/FlinkStream.java:
```diff
@@ -0,0 +1,91 @@
+
+	@Override
+	public boolean canBeInitialised() {
+		return procItem.isInitialised();
```
Is this semantically correct?

senorcarbone replied:
Since there are no named streams in Apache Flink we tag the stream names and filter them out for each consumer task depending on their subscription. This yields the same behavior with Storm's or Samza's named streams and it is still the same performance-wise since filter operators are chained in the same thread.

gdfm commented:
Thanks, good job! In general it looks good. There are a few things that are not clear to me, probably because of my limited knowledge of Flink. I will run some tests to be sure everything works.
We also need another committer to have a look at it. @arinto @abifet could you check it?

On a different note, I'd like to get some feedback on the API.
Was it easy to implement the binding? Was it cumbersome?
Are there parts that are unclear or seem unreasonable?
Would you change something if you could?

gdfm commented:
I am testing it with Flink-0.9-SNAPSHOT from the website (Version: 0.9-SNAPSHOT, Rev:4672e95) but I am getting an exception that looks like I am not using the right version.

```java.lang.NoSuchMethodError: org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.addSource(Lorg/apache/flink/streaming/api/function/source/SourceFunction;Lorg/apache/flink/api/common/typeinfo/TypeInformation;)Lorg/apache/flink/streaming/api/datastream/DataStreamSource;
	at com.yahoo.labs.flink.topology.impl.FlinkEntranceProcessingItem.initialise(FlinkEntranceProcessingItem.java:54)```

senorcarbone commented:
Hi Gianmarco! You should download the binaries of the milestone-1 version [1], the current snapshot will not work.

[1] https://flink.apache.org/downloads.html#preview

gdfm commented:
Thanks, I managed to make it run on Flink.

I am testing the VHT algorithm with ```bin/samoa flink target/SAMOA-Flink-0.3.0-SNAPSHOT.jar "PrequentialEvaluation -d /tmp/dump.csv -i 1000000 -f 100000 -l (classifiers.trees.VerticalHoeffdingTree -p 4) -s (generators.RandomTreeGenerator -c 2 -o 10 -u 10)"```
I had to increase the task manager slots considerably to make it run. I see many "Filter" tasks in Flink's dashboard. Is that normal?
The accuracy of the VHT is reasonable, and comparable with what we get with Storm. It's lower than with the local execution engine, but that's an algorithmic problem we are working on.
I also feel that's a bit slow. I didn't get the output from the console telling me how long it took, but it took a while and the fan of my laptop started spinning. Do you think there is margin for improvement in this respect?

senorcarbone commented:
Hi Gianmarco. To your points:

- We use a filter task per stream subscription so we can forward the right streams into the right task respectively. The filters are basically chained with the previous operators so they do not consume additional slots. 

- Regarding the performance, did you check roughly how much additional time it takes on Flink? We had the opposite experience from our side. I can run VHT with the same settings locally and in EC2 today to double-check and let you know.

- I will also address all style comments as well and update the PR.

Thanks for the good feedback so far!

senorcarbone commented:
Just to clarify slot allocation: A slot is shared by multiple operators and, at the current version at least, the job manager allocates as many slots as the maximum number of parallelism in your streaming job. For example, if you have something like a source->filter->map->filter->sink with parallelism 8 it will allocate 8 slots to run parallel the operator instances of your job.

senorcarbone commented:
I addressed your comments in the last commits. You were right about taking a while to finish, there were some things missing for proper termination but it's fixed now.

senorcarbone commented:
Hey @gdfm @arinto @abifet . Thanks for the review so far. Let me know if you have any additional comments to look into. I think there is nothing critical left, we will keep supporting the flink adapter with subsequent updates such as new releases and improvements.

gdfm commented:
Hi @senorcarbone,
Thanks, I will have another look at it tomorrow.
We also need another review from @abifet or @arinto.
And thanks for offering to keep supporting the adapter :)

abifet commented:
Hi @senorcarbone,

This adapter for Apache Flink seems amazing! Thanks so much! I tested and reviewed, and found the same issues as @gdfm. The accuracy of the VHT was quite good, and it was very slow in my computer. Any thoughts about this? Do we need to tune any parameter in the conf file?

rmetzger commented:
We have a open JIRA in Flink to start a "streaming only" mode, which starts up Flink in an streaming optimized mode.

For now, you can set the following configuration values in the conf/flink-conf.yaml to achieve a similar effect:
- `taskmanager.memory.fraction`: This fraction describes the division between managed and unmanaged heap space. Flink batch is using managed memory for its operations, flink streaming uses unmanaged memory only. By default the value is set to 0.7, which leaves not much memory for streaming. I would recommend setting the value to `0.001` ;)
- `taskmanager.heap.mb`: Thats an integer specifying the task manager's heap size. By default its set to 512. I suspect your machine has more memory.

senorcarbone commented:
Hello again @gdfm and @abifet ,
I did a lot of cross-profiling between storm and flink, running the same `VerticalHoeffdingTree` task under different configurations during the last two days and I think the results are quite interesting. 

It looks like the algorithm performance (and accuracy) depends heavily on the ingestion speed of the local statistics processors. The paradox is that the greater the speed the slower the whole computation gets by time  since more and more attribute events are sent to the local statistics processors with higher rate, the more updates the model aggregator gets back. 

The average processing delay (in num of flatten instances processed by the aggregator between sending a process event and receiving the respective local statistics) is ~2k instances for Flink and around 400k instances for Storm. Also in Storm the aggregator continuously broadcasts ~100-200 attribute messages to local processors on average while Flink broadcasts ~2100 attribute messages due to the rate it gets results back I assume. These are collected locally on each component and there was no message duplication. 
Since you worked on the algorithm, do you find this behavior reasonable?

senorcarbone commented:
I also tried adding a 2sec sleep on the local flink processors to delay their results back and the algorithm finished in 12sec with less (57%) accuracy. Perhaps the model aggregator could be enhanced with some flow control logic to compensate between model updating rates and accuracy.

gdfm commented:
Hi @senorcarbone,
Thanks for digging into the issue.
Interesting findings, though I still need to digest what they mean exactly.
We know that accuracy depends on speed, and the higher the speed the lower the accuracy. What I expect, however, is that the number of messages generated by the algorithm is independent of the speed (at least mostly). If this is not the case, there must be some wrong assumption I am making about the system.

senorcarbone commented:
I think it has to do with the number of splits. If the model aggr. gets more local statistics during the time it processes the same amount of instances, it will have to split (exponentially?) more times and send more attributes in the same period. That is what I got from the experiments at least.  

Maybe there should be a separate issue on VHT but you can keep me and @fobeligi in sync regarding it.  @fobeligi is working on an experimental native implementation of VHT on Flink which we can share soon and she had to deal with similar issues. Also, feel free to let me know if you want me to try out more experiments and share results to speed up the process.

Regarding the PR, do you think we should look into something more?

gdfm commented:
Indeed, I see what you mean. Given that the feedback loop in Flink is faster, the number of attempts to split should increase.
This is expected, but the number of such attempts is upper bounded by the ones tried on the Local engine, where there is no delay between request of the split criterion and response by the local statistics.

We already have some flow control to regulate the rate of ingestion in PrequentialEvaluation. I'll play a bit with it to see what happens.
When you put the 2 seconds delay in the Flink Processors, what happens (I guess) is that the whole data streams through a very rough, sub-optimal version of the tree. So it's very fast, but the precision drops considerably because of the artificial limit on the number of split attempts.

gdfm commented on samoa-instances/src/main/java/com/yahoo/labs/samoa/instances/SingleLabelInstance.java:
```diff
@@ -257,4 +257,28 @@ public String toString() {
+	public void setInstanceInformation(InstancesHeader instanceInformation) {
+		this.instanceInformation = instanceInformation;
+	}
+
```
Why do we need these changes for Flink?

senorcarbone replied:
these were test leftovers apparently, we do need them according to @fobeligi

gdfm commented:
Hi,

I think we should address the issue with VHT in a separate PR.
I'd like to finalize this one.

@senorcarbone there are just a few issues outstanding.
There are a few files in samoa-api/build which should not be part of the patch.
Also, there is a question about the partitioning enum that I'd like to be answered.
Finally, there are some changes in samoa-instances that are not clear to me (I had missed them before).

Once these are fixed I'm +1

senorcarbone commented:
Thanks @gdfm . I hope I addressed everything in the last commit and inline comments. Let me know if you want me to look into something more!

gdfm commented:
Yes, the patch looks good to me.
+1

@abifet @arinto any comment?
I will merge the patch tomorrow otherwise.

abifet commented:
+1

gdfm commented:
Merged.
@senorcarbone thanks for the contribution!

rmetzger commented:
Great to see this merged.

Is the Samoa website located in the `gh-pages` branch?
I would like to update the front page to mention Flink as well ;)

gdfm commented:
Yes, the sources are in gh-pages and we use jekyll to build the website.
@rmetzger it would be great if you could add docs :)
Instructions here: https://cwiki.apache.org/confluence/display/SAMOA/SAMOA+Home

rmetzger commented:
Thank you.
I've filed a JIRA: https://issues.apache.org/jira/browse/SAMOA-33
Lets see when I have time for this ;)

senorcarbone commented:
Awesome, thanks @gdfm! Will you merge from local [branch](https://github.com/gdfm/incubator-samoa/tree/SAMOA-16)? Mind that it is missing some last commits. I will close the PR once everything is merged.

On a slightly related note, we ran some experiments today on the new stable streaming api that will be released soon that contains many good fixes and additions. The VHT experiment mentioned above now takes approximately around 80sec for 1m instances since filters are properly chained, making input ingestion much faster so that it catches up with local processors. You can find the branch with the patch [here](https://github.com/senorcarbone/incubator-samoa/tree/releasepatch). Special thanks go to @gyfora for fixing chaining in iterations!  

I will create a new PR once we have a stable release.

gdfm commented:
Weird, I have already merged the PR and pushed back to Apache.
https://git-wip-us.apache.org/repos/asf?p=incubator-samoa.git

Something in the mirroring is not working, the PR should be closed automatically.

Thanks for the info, improving performances would be a great contribution!
